library/core/src/slice/sort.rs

```diff
@@ -27,10 +29,27 @@ impl<T> Drop for CopyOnDrop<T> {
     }
 }
 
+const fn lt<T>(lhs: *const T, rhs: *const T) -> bool {
+    let zero = 0;
+    // SAFETY:  This is a helper function.
+    //          Please refer to `offset_from` for correct usage.
+    //          Mainly lhs and rhs being in the same allocated Object
+    //            and a multiple of their size apart.
+    unsafe { lhs.offset_from(rhs) }.lt(&zero)
+}
+const fn eq<T>(lhs: *const T, rhs: *const T) -> bool {
+    let zero = 0;
+    // SAFETY:  This is a helper function.
+    //          Please refer to `offset_from` for correct usage.
+    //          Mainly lhs and rhs being in the same allocated Object
+    //            and a multiple of their size apart.
+    unsafe { lhs.offset_from(rhs) }.eq(&zero)
+}
```

Comment:
Should those helper functions themself be unsafe?

Reply:
Yes they should, and they should document their precondition.

Reply:
That indicates that the code is repeatedly doing the same subtle safety-relevant reasoning. This doesn't get better by not talking about it, it only gets better by restructuring the code in a way that avoids doing subtle safety-relevant reasoning in many places. (I don't know if that is possible, I don't have the time to do a detailed code review -- sorry.)